### ISSUE TITLE
causality and variability attribute

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#33]
**Reported by torstenblochwitz on 6 Jan 2010 08:14 UTC**
I wonder why we have chosen parameter as variability and not as causality attribute. May be I forgot the reason.
If we would use it as causality attribut (which would be more logical for me) we could use the variability attribute to indicate parameters which can be changed after fmiInitialize in a later version.

----

Migrated-From: https://trac.fmi-standard.org/ticket/33

**Comment by modelica-trac-importer** [comment]
**Comment by jakob on 12 Jan 2010 10:34 UTC**
Actually, there are 3 different sets of parameters of a model:
 * user parameter: parameter declared as such by a user e.g. using Modelica. A code generator
   may actually turn this into a constant, e.g. for a parameter that specifies an array size (structural parameter) or if unable to generate code that works for all possible parameter values.
 * code parameter: can be set before model init
 * tuneable parameter: can be set to a new value during simulation
The XML file does not currently distinguish these cases, and variability=parameter means "code parameter".

Attributes in current FMI draft:
 * causality: input output internal none
 * variability: constant parameter discrete continuous
Proposal:
 * causality: input output **parameter** internal none
 * variability: as before
where a pair (causality/variability) 
 * parameter/constant: means "user parameter" turned into constant by code genrator
 * parameter/parameter: means "user parameter" turned into a "code parameter"
 * parameter/discrete: means "user parameter" turned into a "tuneable parameter"

**Comment by modelica-trac-importer** [comment]
**Comment by jakob on 12 Jan 2010 11:30 UTC**
The 3 cases above cover "independent parameters" only. 
A dependent parameter could be represented by the pair 
 * causality = internal
 * variability = maximum of the variabilities of indepent parameters used to compute the dependent parameter.
Unfortunately, we loose the info that the dependent parameter is also a user parameter. A fix is to use causality = dependentParameter for this case.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 20 Jan 2010 13:32 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 7 Apr 2011 09:42 UTC**
The meaning of the attributes is:

- Varability: Defines when the value of a variable changes (constant: does not change; parameter: does not change after initialization, discrete: only changes at events; continuous: no restrictions).

- Causality: Defines how the variable is visible from the outside of the model and defines when this information is needed when the FMU is connected to other FMUs.

So, to summarize, these two attributes are completely orthogonal and should not be mixed.

Furthermore, there are two types of parameters: If in the xml file a "start" attributes is set for a parameter, the parameter can be changed before simulation starts ("independent parameter"). If this is not the case, this is a "dependent parameter" that is computed from "independent parameters".

"Tunable parameter" means to

1. Stop the simulation at a step event.
2. Change the value of the parameter
3. Newly start the simulation using as initial values the actual values of all previous variables and the new value of the parameters. Changing a parameter, requires to newly compute all dependent parameters.

There is the strong goal to introduce "tunable parameters" in FMI 2.0. There seems to be the following variants:

There are the following issues:


**Proposal 1** (from Hilding at the last FMI meeting):

All (independent) parameters are tuneable, i.e., variables that have `variability = parameter` and have a `start value `

Advantage: Simple.

Disadvantages: 

- Not all independent parameters might be tunable, for example it does not make sense to tune the start/initial value of a variable that is defined by a variable.

- Tunable independent parameters, and parameters that depend on these parameters, can change during simulation. This means that they must also be stored in the result file. Therefore, with this proposal all parameters must be stored at events, although in most cases, the parameter never changes and could be stored just once. Since a model might have potentially many parameters, this could give a real performance degragation.


"""Proposal 2""" (from Martin)

1. Variability is changed to "constant, parameter, tunableParameter, discrete, continuous". All independent tunableParameters must have a start value. A tunableParameter without a start value is a dependent tunableParameter.

2. The capability flag "completedIntegratorStepNotNeeded" must be set to true, if parameters can be tuned (otherwise no step events can occur).

A tool would work in the following way:

1. If the user changes a parameter, then the tool has to wait until the next integrator step is completed.

2. Function `fmiCompletedIntegratorStep` must be called

3. Via the `fmiSetXXX` functions, new independent tunable parameters are set.

4. Function `fmiEventUpdate` must be called.

Advantage. Can be efficiently handled: At event times, tunable parameters are also stored in the result file. All other parameters are stored only once.

Disadvantage: Schema/description becomes more complicated.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 17 Apr 2011 20:45 UTC**
Attached is a good proposal to improve "variable categories" and replace the enumerations "Variability" and "Causality". The proposal is from Jakob Mauss with some adaptations from me (Martin Otter)

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 19 Apr 2011 12:47 UTC**
I don't have any strong objections to that proposal, but it does not look like orthogonal categories to me; and thus even though there are many categories it does not cover all cases (more at the end).

I believe that with a slight change from the previous draft we could handle it in a nice orthogonal way. I would propose three categories:
Variability: constant, nontuneableParameter, parameter, discrete, continuous
Causality: input, output, internal, none
originalVariability: constant, (nontuneableParameter) parameter, discrete, continuous (this is only intended to help users by indicating how how the variable was declared originally; it could be made optional; preferably with capital initial letter but I have wiki-problems with OriginalVariability)

The combination Variability=constant and Causality=input is illegal, the rest ok.
Furthermore it is required that Variability<=originalVariability

fmiSet is only allowed if causality is input or none

Definition of variability:
* constant: Does not change
* nontuneableParameter: Can only be set before initialization
* parameter: Only changing at certain event points where fmiSet is called for an "input parameter"; i.e. the model does not change a parameter on its own
* discrete: Only changing at events
* continuous: Changing all the time

The benefit of making it orthogonal is that we can handle:
* Modelica.Blocks.Sources.Constant. The output from this one should be marked as Variability=parameter (or if evaluated 'constant'), Causality=output, originalVariability=continuous. With the proposal above we cannot mark an output as being parametric.
* Tool specific variables that should only be set at the start of the simulation.
* Only storing dependent parameters once if they do not depend on any tuneable parameter.

Note: The reason for calling it nontuneableParameter and parameter instead of parameter and tuneableParameter is that dependent parameters can be marked with either of these as well, and the safe approach is to consider them as tuneable parameters - which should thus have a short name.

**Comment by modelica-trac-importer** [comment]
**Comment by TorstenBlochwitz on 19 Apr 2011 15:17 UTC**
I like Jakobs and Martins proposal in principle. I added some remarks (see attachment VariableCategory_v3_ITI.doc).
What I do not like is the property **wasParameter** (see attachement) and the categories **constant** and **dependentParameter** (see below).

For me constants and dependent parameters are internal variables (category: **internal**). (If the user defines it during FMU generation they can be outputs too.) They do have special properties, that indicate when they might be changed by the FMU: 
  * never (new property: **constant**)
  * after a tunable parameter was changed (property: **tunable**).
 
This would also handle Hans example (output of Modelica.Blocks.Sources.Constant).

May be we do come to an orthogonal definition if we distiguish the properties for variables that can be set by the environment (inputs and parameters) and variables that can be retreived only (internals and outputs).
For settable variables we have to define at which state fmiSetXXX is allowed to be called:
   * only before fmiInitialize for parameters,
   * at events for discrete inputs and tunable parameters,
   * always for continuous inputs.

(By the way tunable parameters are a kind of discrete: they can be changed at events only. The event is not triggered by time or eventIndicators, but by the environment. At the end the event might be triggered by the enclosing model too: e.g. depending on something the parameter gain of a PID controller (implemented as FMU) is switched to another value by the enclosing model to simulate a simple adaptive controller).

For "only gettable" variables (outputs and internals) we have to define, when they might change inside the FMU:
   * always (continuous)
   * at events (discrete)
   * never after initialization (constant),
   * after a parameter change ("special discrete" or discrete only).
This allows an optimization of the calling sequence: it defines when an fmiGetXXX call might make sense.

The other classification (input vs. parameter, internal vs. output) eases the embedding of the FMU into a tool and the usability:
   * e.g. parameters are shown in a properties window, for inputs not
   * e.g. for inputs signal inputs are generated (in tools like Simulink)
   * e.g. it is allowed to connect other inputs to an output (not to an internal)

On the other hand there are also some mathematical classifications:
   * e.g. a FMU can imply dp/dt = 0 for parameters p
   * e.g. the dependency of internals w.r.t. inputs is not defined by the FMU
   * e.g. the Jacobian of internals w.r.t anything can not be provided by the FMU

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 19 Apr 2011 15:28 UTC**
Here is a respond to Hans submission (not taking into account the submission from Torsten):

It is not completely clear how dependent parameters are defined. It could be:
- parameter without a start value
- parameter with causality = output (see below)
This should be clarified.

The issue with a (Modelica) input signal that is turned into a constant during compilation needs more discussion:

- From a formal point of view, an FMU does not define any connections. The only thing that an FMU knows is that variables can be "set" with the "fmiSetXXX" functions and can be get with "fmiGetXXX" functions. Saying it differently, only input and output arguments are present.

- In the proposal from Jakob and me, the logic is that "input discrete" and "input continuous" is time variying, whereas "constant" and "parameter" is not. Therefore, an input signal that is turned into a constant (probably a parameter), would be defined as "parameter" in the FMU.

The idea with "OriginalVariability" is good.

I discussed with Jakob also another proposal that is better from a formal point of view, but is less understandable (and therefore I did not send it previously):


```
Enumeration: Environment:
   fromEnvironment: Provided from the environment and utilized by FMU
   toEnvironment  : Computed inside the FMU and utilized by other
                    models in the environment.
   internal       : Computed inside the FMU and cannot be utilized by
                    other models in the environment

Enumeration: TimeDependency
   constant        : The value of the variable is fixed and never changes
   parameter       : The value of the variable is constant after initialization
   tunableParameter: The value of the variable is constant after initialization.
                     The value can be changed at events, leading to a
                     re-initialization.
   discrete        : The value of the variable can be changed at events
   continuous      : The value of the variable is continuous between events
```

This proposal is very clean with orthogonal definitions, but it is more difficult to understand certain (useful) combinations:

```
fromEnvironment - parameter:  independent parameter that cannot be tuned.
toEnvironment   - parameter:  dependent parameter that depends on
                              untunable, independent parameter
```

This proposal is very similar to the proposal from Hans with just different naming:

- input -> fromEnvironment (because in the FMU function interface there is notion of "input signal".
- output -> toEnvironment (same argument as for "input").
- nontunableParameter -> parameter (parameter/tunableParameter looks nicer as nontunableParameter/parameter)

Still I think that the proposal from Jakob and me is better, because it is easier to grasp when one sees the definitions (however, I can also live with this new proposal and the proposal from Hans, provided "output parameter" means "dependent parameter").

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 19 Apr 2011 16:31 UTC**
Lets add one more issue:

  When importing an FMU in to Modelica (and in any other environment with an object diagram editor), it is unclear how the inputs and outputs should be represented. If there is more as one input or more as one output, this is very annoying. The tool has several options, e.g., display every scalar input as separate inputs. But, if there are more then 3 or 4 inputs, this does no longer look nice. Another option in Modelica would be to define an "expandable connector" and store all inputs and output in it (or have two of them).

So, clearly not all information of the model from which the FMU was generated, can be passed to the FMU and therefore importing an FMU does not retain the original model interface.

The remedy is to store more information about the connector definition of the original model in the FMU. This could be performed with a vector of additional elements, such as:

```
  <connectors>
    <connector name="Flange_a", placement="xxxx">
      <signal name="abc.cde.phi">
      <signal name="abc.cde.torque", flow=true>
    </connector>
  </connectors>
```

If this approach is used, then the attributes of a signal should clearly only define the causality ("is provided FROM the environment", and "is provided TO the environment"). If a parameter or constant is used in a <connector> definition, then clearly, it is in the connection and not in a parameter menu (currently it is not possible to propagate a parameter via a Modelica connector, but this does not mean that this definition is not useful).

So, to summarize, it seems more general to have the following views:

- Causality, i.e., in which part of the model hierarchy is the variable computed.
- Time-dependability, i.e, at which points in time is the variable computed.
- Environment-usage, i.e., how is the expected usage of the variable in the environment.

In Beta 1, we could skip the "Environment-usage" since not enough time to define it. However, when we want to introduce it later, then we should not introduce "Environment-usage" implicitely in the current attributes (which means that the proposal from Jakob and me seems not be good in this respect).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 Apr 2011 07:25 UTC**
We had missed one issue and then the rest follows logically:

 We have the "direct term",or say it in other ways, need to define Jacobians. In both cases the variables must be clearly identified that can communicate with external models. The most important underlying issue is whether a variable is coming from another model and is used in another model, because then algebraic loops might occur when connecting the FMU, and this must be handled. The consequence is that we must introduce a category for variables that are set from the outside, but cannot be used in a model connection. 

As a result, the following categories **must** be used (note, the categories are important; if you do not like the names, we can use different ones):

```
input      : Can be set from a model in the environment.
output     : Can be used from a model in the envionment.
interactive: Can be set from the environment, but not from a model
             (therefore, algebraic loops between models cannot occur and
             no Jacobian for this variable is needed for simulation, but
             optionally for optimization.)
             Alternative name from Hilding: environment.
internal   : Is set in the FMU and cannot be used in a model of the environment
none       : Tool specific (this is not clear to me and should better specified,
             ev. also use a different name, like "vendorSpecific")
```

Now, we need to define at which time instants the fmiGetXXX and fmiSetXXX functions are allowed to be called. This means, we need to define the time variability. We can use the current variability enumeration and adding one for "tunable". There seems to be agreement on the categories, but not on the names:


```
Literal value                              : constant
Can be changed before initialization       : parameter, nontunableParameter,
                                             nontunable, fixed 
Can be changed at events (but constant)    : parameter, tunableParameter, 
                                             tunable
Can be changed at events (but time varying): discrete
No restriction                             : continuous
```

Finally, we have to decide whether all combinations of causality and of variability makes sense.

Hilding pointed out one restriction: At an event instant, the model can be reinitalized and a new simulation can be started. This can only be defined with a restricted form of solving an initialization problem where only changes in parameters are allowed. He proposes that also states can be changed. There is then, however, the problem that this is not well defined due to dynamic state selection, association of start values with parameters and changes of states in a model by impulses. I do not think that we manage to come up with a clean proposal until Thursday for this.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2011 11:14 UTC**
This ticket has been finalized at the FMI meeting on May 26 in Lund and is now incorporated into the FMI 2.0 draft specification. Final solution:

- Causality  : parameter, input, output, internal, option
- Variability: constant, fixed, tunable, discrete, continuous

The proposal to have also an "original variability" was shortly discussed without final conclusion. The current view is:

In a Modelica tool, usually parameters and constants are listed in the variable browser (before simulation starts). Parameters that are not modifiable, are listed here too, but without input field. If a parameter is fixed during compilation, it should be converted to a `constant` and should be within the xml-file. Then, the view of a tool with variable browser is exactly as in the original tool: The constants are shown, but without an input field.

What a user would really like to know is when a tool converts a parameter to a constant and the reason for it. However, this information can only be given in the tool that generates the FMU and at this stage this information should be given.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 19 Mar 2012 10:01 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 7 Sep 2012 10:18 UTC**
